File: day1/day1.py
import scipy.ndimage as sp, numpy as np #np arrays are just nice in general and scipy.ndimage correlate and convolve, rot90 default axis=1 rotates left axis=0 would be right
import re #re.split(), re.match, re.findall, Regex Guide at https://www.debuggex.com/cheatsheet/regex/python
import collections #defaultdicts, maybe counters
import functools #use @functools.lru_cache(None) above a function to keep track of all inputs and speed it up, reduce
from math import log
from string import ascii_uppercase, ascii_lowercase, ascii_letters #ascii_uppercase, ascii_lowercase, etc..
import itertools #itertools.product(range(a),range(b),[1,2,3]) allows for nested for loops in one line essentially
from more_itertools import chunked, windowed #chunked, windowed
from scanf import scanf
import networkx #Networkx for graph problems like connected components
from copy import deepcopy #deepcopy() let's use copy a list without reference areas
import portion # Interval stuff
from z3 import Solver, Int, And
#Sets are {}, A|B finds union of sets, A&B finds intersection, A-B finds difference, A^B is (A|B)-(A&B)
#Python has complex numbers of the form x+yj (1+1j) or complex(x,y) (complex(1,1))

def runPart1(filename):    
    lines = open(filename).read().splitlines()
    sum = 0
    for line in lines:
        res = re.findall(r'\d', line)
        sum += int(res[0]+res[-1])
    return sum        

def runPart2(filename):
    lines = open(filename).read().splitlines()
    sum = 0
    rep = {'one':'1','two':'2','three':'3','four':'4','five':'5','six':'6','seven':'7','eight':'8','nine':'9'}
    for line in lines:
        res = re.findall(rf'\d|{"|".join(rep.keys())}', line)
        res = [rep[x] if x in rep else x for x in res]
        sum += int(res[0]+res[-1])
    return sum

print("Running example.txt...")
print("Part 1:", runPart1("example.txt"))
print("Part 2:", runPart2("example2.txt"))
print()
print("Running input.txt...")
print("Part 1:", runPart1("input.txt"))
print("Part 2:", runPart2("input.txt"))

File: day1/day1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
using namespace std;

int main(int argc, const char** argv) 
{
    int p1_ans, p2_ans;

    // Get input in ifstream
    ifstream file("input.txt");
    if (!file) {
        cout << "Failed to open input.txt" << endl;
    }

    // Map of words to integers
    unordered_map<string, string> hashmap;
    hashmap["one"] = "o1e";
    hashmap["two"] = "t2o";
    hashmap["three"] = "t3e";
    hashmap["four"] = "4";
    hashmap["five"] = "5e";
    hashmap["six"] = "6";
    hashmap["seven"] = "7n";
    hashmap["eight"] = "e8t";
    hashmap["nine"] = "9";

    // Parse file line by line
    string str;
    while (getline(file, str)) {
        
        // Part 1: Just look for integer from start and end of string
        auto it = str.begin();
        while (it != str.end() && !isdigit(*it)) ++it;
        p1_ans += (*it-'0')*10;
        it = str.end()-1;
        while (it != str.begin()-1 && !isdigit(*it)) --it;
        p1_ans += *it-'0';

        // Part 2: Replace one with o1e and so on to get cases
        for (auto entry : hashmap) {
            int start_pos;
            while ((start_pos = str.find(entry.first)) != -1) {
                str.replace(start_pos, entry.first.length(), entry.second);
            }
        }
        it = str.begin();
        while (it != str.end() && !isdigit(*it)) ++it;
        p2_ans += (*it-'0')*10;
        it = str.end()-1;
        while (it != str.begin()-1 && !isdigit(*it)) --it;
        p2_ans += *it-'0';
    }

    cout << "Part 1: " << p1_ans << endl;
    cout << "Part 2: " << p2_ans << endl;
    return 0;
}
